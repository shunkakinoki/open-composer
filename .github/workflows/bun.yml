name: Bun
on:
  push:
    branches:
      - main
  pull_request:
    branches:
      - main
  workflow_dispatch:
concurrency:
  group: ${{ github.workflow }}-${{ github.event_name }}-${{ github.head_ref || github.sha }}
  cancel-in-progress: true
jobs:
  bun-build:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Build
        run: bun run build
  bun-lint:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Check
        run: bun run check
      - name: Format
        run: bun run format
      - name: Lint
        run: bun run lint
      - name: Type Check
        run: bun run type-check
  bun-test:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Test
        run: bun run test
  bun-check:
    if: always()
    needs:
      - bun-build
      - bun-lint
      - bun-test
    runs-on: ubuntu-latest
    timeout-minutes: 3
    steps:
      - name: Alls Green
        uses: re-actors/alls-green@release/v1
        with:
          jobs: ${{ toJSON(needs) }}
