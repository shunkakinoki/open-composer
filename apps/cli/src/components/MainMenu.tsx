import { TextAttributes } from "@opentui/core";

import { useKeyboard } from "@opentui/react"; 
import type React from "react";
import { useState } from "react";

export interface MenuItem {
  key: string;
  label: string;
  description: string;
  onSelect: () => void;
}

interface MainMenuProps {
  items: MenuItem[];
  onExit: () => void;
}

export const MainMenu: React.FC<MainMenuProps> = ({ items, onExit }) => {
  const [selectedIndex, setSelectedIndex] = useState(0);

  useKeyboard((key) => {
    if (key.name === "up" || key.sequence === "k") {
      // Move up
      setSelectedIndex((prev) => (prev > 0 ? prev - 1 : items.length - 1));
    } else if (key.name === "down" || key.sequence === "j") {
      // Move down
      setSelectedIndex((prev) => (prev < items.length - 1 ? prev + 1 : 0));
    } else if (key.name === "return") {
      // Select current item
      items[selectedIndex]?.onSelect();
    } else if (key.sequence === "q" || key.name === "escape") {
      // Exit
      onExit();
    } else if (key.sequence && key.sequence >= "1" && key.sequence <= "9") {
      // Quick select by number
      const index = Number.parseInt(key.sequence, 10) - 1;
      if (index >= 0 && index < items.length) {
        items[index]?.onSelect();
      }
    }
  });

  return (
    <box flexDirection="column" padding={1}>
      <text content="Main Menu" style={{ fg: "cyan", attributes: TextAttributes.BOLD }} />
      <box borderStyle="single" borderColor="gray" marginTop={1} />

      <box flexDirection="column" marginTop={1}>
        {items.map((item, index) => {
          const isSelected = index === selectedIndex;
          const numberLabel = `[${index + 1}]`;
          const prefix = `${isSelected ? "▶ " : "  "}${numberLabel.padEnd(4)}`;

          return (
            <box key={item.key} marginY={0}>
              <text
                content={prefix}
                style={{ fg: isSelected ? "green" : "gray" }}
              />
              <text
                content={item.label.padEnd(16)}
                style={{
                  fg: isSelected ? "cyan" : "white",
                  attributes: isSelected ? TextAttributes.BOLD : undefined
                }}
              />
              <text content={item.description} style={{ fg: "gray" }} />
            </box>
          );
        })}
      </box>

      <box marginTop={2}>
        <text content="Use ↑↓/j/k to navigate, Enter to select, 'q' to quit" style={{ fg: "gray" }} />
      </box>
    </box>
  );
};
