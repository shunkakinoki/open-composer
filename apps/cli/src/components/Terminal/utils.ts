/**
 * Terminal utility functions
 * Inspired by termact: https://github.com/MasterGordon/termact/blob/main/src/utils.ts
 */

import type { TerminalStyle } from "./types.js";
import {
  bg,
  blink,
  bold,
  doubleUnderline,
  fg,
  hidden,
  inverse,
  italic,
  strikethrough,
  underline,
} from "./ansi.js";

export const applyStyle = (style: TerminalStyle): string => {
  let result = "";
  if (style.fg) result += fg(style.fg);
  if (style.bg) result += bg(style.bg);
  if (style.bold) result += bold();
  if (style.italic) result += italic();
  if (style.underline) result += underline();
  if (style.blink) result += blink();
  if (style.reverse) result += inverse();
  if (style.hidden) result += hidden();
  if (style.strikethrough) result += strikethrough();
  if (style.doubleUnderline) result += doubleUnderline();
  return result;
};

export const wrapText = (text: string, width: number): string[] => {
  const lines: string[] = [];
  const words = text.split(" ");
  let currentLine = "";

  for (const word of words) {
    if (currentLine.length + word.length + 1 <= width) {
      currentLine += (currentLine ? " " : "") + word;
    } else {
      if (currentLine) lines.push(currentLine);
      currentLine = word;
    }
  }

  if (currentLine) lines.push(currentLine);
  return lines;
};

export const truncateText = (text: string, maxLength: number): string => {
  if (text.length <= maxLength) {
    return text;
  }
  return text.slice(0, maxLength - 3) + "...";
};

export const padText = (
  text: string,
  width: number,
  align: "left" | "center" | "right" = "left",
): string => {
  if (text.length >= width) {
    return text.slice(0, width);
  }

  const padding = width - text.length;

  switch (align) {
    case "center": {
      const leftPad = Math.floor(padding / 2);
      const rightPad = padding - leftPad;
      return " ".repeat(leftPad) + text + " ".repeat(rightPad);
    }
    case "right":
      return " ".repeat(padding) + text;
    default:
      return text + " ".repeat(padding);
  }
};

export const getNtscGrayscale = (r: number, g: number, b: number): number => {
  return r * 0.299 + g * 0.587 + b * 0.114;
};
