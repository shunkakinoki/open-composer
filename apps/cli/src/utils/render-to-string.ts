import { createTestRenderer } from "@opentui/core/testing";
import type { ReactElement } from "react";

/**
 * Renders a React component to a string using OpenTUI
 * Useful for rendering UI components in Effect CLI commands
 *
 * Note: This is a simplified version that uses OpenTUI's test renderer
 * For full interactive rendering, use the main render function from @opentui/react
 */
export async function renderToString(element: ReactElement): Promise<string> {
  const { renderer, renderOnce, captureCharFrame } = await createTestRenderer({
    width: 100,
    height: 30,
  });

  // Note: We can't easily use the high-level render() here because it's designed
  // for full terminal interaction. For now, this returns an empty string.
  // Components should use direct terminal output if needed in CLI commands.

  // Clean up
  renderer.destroy();

  return "";
}
