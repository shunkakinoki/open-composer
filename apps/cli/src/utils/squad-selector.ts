import inquirer from "inquirer";
import type { LLMTask } from "@open-composer/agent-registry";
import { getAgentRegistry, getSquadLauncher } from "@open-composer/agent-registry";

/**
 * Start interactive squad selector
 */
export async function startSquadSelector(): Promise<void> {
  const registry = getAgentRegistry();
  const launcher = getSquadLauncher();

  console.log("\n🎮 Pokemon-Style Squad Manager\n");

  const { action } = await inquirer.prompt([
    {
      type: "list",
      name: "action",
      message: "What would you like to do?",
      choices: [
        { name: "📋 List all agents", value: "list" },
        { name: "✨ Create new squad", value: "create" },
        { name: "🚀 Quick squad (by task)", value: "quick" },
        { name: "📊 View statistics", value: "stats" },
        { name: "🚪 Exit", value: "exit" },
      ],
    },
  ]);

  switch (action) {
    case "list": {
      const agents = registry.getAllAgents();
      console.log(`\n📦 Found ${agents.length} agents:\n`);
      for (const agent of agents) {
        console.log(
          `  ${agent.pokemonAttributes.sprite} ${agent.name} (${agent.tier}) - ${agent.provider}`,
        );
      }
      break;
    }

    case "create": {
      console.log("\n🛠️  Create Squad - Not implemented in CLI yet");
      console.log("   Use the interactive UI components for full squad creation");
      break;
    }

    case "quick": {
      const { task } = await inquirer.prompt([
        {
          type: "list",
          name: "task",
          message: "Select task specialization:",
          choices: [
            { name: "🧭 Planner", value: "planner" },
            { name: "⚡ Programmer", value: "programmer" },
            { name: "🔍 Reviewer", value: "reviewer" },
            { name: "🌊 Router", value: "router" },
            { name: "🌿 Summarizer", value: "summarizer" },
          ],
        },
      ]);

      const team = launcher.createQuickSquad(task as LLMTask);
      console.log(`\n✅ Created squad: ${team.name}`);
      console.log(`   Agents: ${team.agents.map((a) => a.name).join(", ")}`);
      break;
    }

    case "stats": {
      const stats = registry.getStats();
      console.log("\n📊 Registry Statistics:");
      console.log(`   Total Agents: ${stats.totalAgents}`);
      console.log(`   Total Squads: ${stats.totalTeams}`);
      console.log(`   Starter: ${stats.tierCounts.starter || 0}`);
      console.log(`   Evolved: ${stats.tierCounts.evolved || 0}`);
      console.log(`   Legendary: ${stats.tierCounts.legendary || 0}`);
      break;
    }

    case "exit":
      console.log("\n👋 Goodbye!\n");
      return;
  }

  // Loop back to main menu
  await startSquadSelector();
}
