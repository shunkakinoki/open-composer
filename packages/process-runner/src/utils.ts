import * as Effect from "effect/Effect";
import {
  MAX_COMMAND_LENGTH,
  MAX_RUN_NAME_LENGTH,
  RUN_NAME_PATTERN,
} from "./constants.js";
import { ProcessRunnerError } from "./types.js";

// =============================================================================
// VALIDATION UTILITIES
// =============================================================================

export function validateRunName(
  runName: string,
): Effect.Effect<string, ProcessRunnerError> {
  if (!runName || typeof runName !== "string") {
    return Effect.fail(
      ProcessRunnerError("Run name must be a non-empty string"),
    );
  }

  if (runName.length > MAX_RUN_NAME_LENGTH) {
    return Effect.fail(
      ProcessRunnerError(
        `Run name too long (max ${MAX_RUN_NAME_LENGTH} characters)`,
      ),
    );
  }

  if (!RUN_NAME_PATTERN.test(runName)) {
    return Effect.fail(
      ProcessRunnerError(
        "Run name can only contain letters, numbers, hyphens, and underscores",
      ),
    );
  }

  return Effect.succeed(runName);
}

export function validateCommand(
  command: string,
): Effect.Effect<string, ProcessRunnerError> {
  if (!command || typeof command !== "string") {
    return Effect.fail(
      ProcessRunnerError("Command must be a non-empty string"),
    );
  }

  if (command.length > MAX_COMMAND_LENGTH) {
    return Effect.fail(
      ProcessRunnerError(
        `Command too long (max ${MAX_COMMAND_LENGTH} characters)`,
      ),
    );
  }

  // Basic command injection prevention - only reject null bytes
  if (command.includes("\x00")) {
    return Effect.fail(
      ProcessRunnerError("Command contains invalid characters"),
    );
  }

  return Effect.succeed(command);
}

// =============================================================================
// UTILITY FUNCTIONS
// =============================================================================

export function withTimeout<T>(
  effect: Effect.Effect<T, ProcessRunnerError>,
  timeoutMs: number,
  timeoutMessage: string = "Operation timed out",
): Effect.Effect<T, ProcessRunnerError> {
  const timeoutEffect = Effect.async<T, ProcessRunnerError>((resume) => {
    const timeout = setTimeout(() => {
      resume(Effect.fail(ProcessRunnerError(timeoutMessage)));
    }, timeoutMs);
    return Effect.sync(() => clearTimeout(timeout));
  });

  return Effect.race(effect, timeoutEffect);
}
