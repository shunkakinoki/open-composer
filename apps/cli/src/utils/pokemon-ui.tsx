import type { AgentTeam, OpenComposerAgent } from "@open-composer/agent-registry";
import type { RegistryStats } from "@open-composer/analytics";
import { renderToString } from "./render-to-string.js";
import {
  AgentCard,
  AgentComparison,
  AgentList,
  EvolutionTree,
  RegistryStats as RegistryStatsComponent,
  TaskSpecializations,
  TeamDisplay,
} from "../components/PokemonUI.js";

/**
 * Pokemon-style UI renderer using React components
 * This is a utility wrapper that renders React components to strings for CLI output
 */
export class PokemonUI {
  /**
   * Render agent card
   */
  static renderAgentCard(agent: OpenComposerAgent): string {
    return renderToString(<AgentCard agent={agent} />);
  }

  /**
   * Render agent list
   */
  static renderAgentList(agents: OpenComposerAgent[]): string {
    return renderToString(<AgentList agents={agents} />);
  }

  /**
   * Render team display
   */
  static renderTeam(team: AgentTeam): string {
    return renderToString(<TeamDisplay team={team} />);
  }

  /**
   * Render evolution tree
   */
  static renderEvolutionTree(agent: OpenComposerAgent): string {
    return renderToString(<EvolutionTree agent={agent} />);
  }

  /**
   * Render task specializations
   */
  static renderTaskSpecializations(agent: OpenComposerAgent): string {
    return renderToString(<TaskSpecializations agent={agent} />);
  }

  /**
   * Render registry statistics
   */
  static renderStats(stats: RegistryStats): string {
    return renderToString(<RegistryStatsComponent stats={stats} />);
  }

  /**
   * Render agent comparison
   */
  static renderComparison(
    agent1: OpenComposerAgent,
    agent2: OpenComposerAgent,
  ): string {
    return renderToString(<AgentComparison agent1={agent1} agent2={agent2} />);
  }
}
