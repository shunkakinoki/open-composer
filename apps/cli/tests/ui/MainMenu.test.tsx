import { expect, test } from "bun:test";
import { MainMenu, type MenuItem } from "../../src/components/MainMenu.js";
import { render } from "../utils.js";

test("MainMenu renders menu items", () => {
  const mockItems: MenuItem[] = [
    {
      key: "test1",
      label: "Test 1",
      description: "First test item",
      onSelect: () => {},
    },
    {
      key: "test2",
      label: "Test 2",
      description: "Second test item",
      onSelect: () => {},
    },
  ];

  const { lastFrame } = render(
    <MainMenu items={mockItems} onExit={() => {}} />,
  );

  const output = lastFrame();

  // Check that menu title is present
  expect(output).toContain("Main Menu");

  // Check that menu items are rendered
  expect(output).toContain("Test 1");
  expect(output).toContain("First test item");
  expect(output).toContain("Test 2");
  expect(output).toContain("Second test item");

  // Check that navigation hints are present
  expect(output).toContain("Use ↑↓/j/k to navigate");
});

test("MainMenu renders with numbered items", () => {
  const mockItems: MenuItem[] = [
    {
      key: "first",
      label: "First",
      description: "First item",
      onSelect: () => {},
    },
  ];

  const { lastFrame } = render(
    <MainMenu items={mockItems} onExit={() => {}} />,
  );

  const output = lastFrame();

  // Check that item has number label
  expect(output).toContain("[1]");
});

test("MainMenu matches snapshot", () => {
  const mockItems: MenuItem[] = [
    {
      key: "sessions",
      label: "Sessions",
      description: "List all sessions",
      onSelect: () => {},
    },
    {
      key: "cache",
      label: "Cache",
      description: "Manage cache",
      onSelect: () => {},
    },
    {
      key: "worktree",
      label: "Worktree",
      description: "Manage worktrees",
      onSelect: () => {},
    },
  ];

  const { lastFrame } = render(
    <MainMenu items={mockItems} onExit={() => {}} />,
  );

  expect(lastFrame()).toMatchSnapshot();
});
