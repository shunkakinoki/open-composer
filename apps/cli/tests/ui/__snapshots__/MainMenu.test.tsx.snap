// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`MainMenu matches snapshot 1`] = `
"                                                                                
 Main Menu                                                                      
                                                                                
 ┌────────────────────────────────────────────────────────────────────────────┐ 
 └────────────────────────────────────────────────────────────────────────────┘ 
                                                                                
 ▶ [1]                                                                          
 Sessions                                                                       
 List all sessions                                                              
   [2]                                                                          
 Cache                                                                          
 Manage cache                                                                   
   [3]                                                                          
 Worktree                                                                       
 Manage worktrees                                                               
                                                                                
                                                                                
 Use ↑↓/j/k to navigate, Enter to select, 'q' to quit                           
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
"
`;
